snapshots/windows/common.go

```diff
@@ -0,0 +1,419 @@
+		}
+
+		path = w.getSnapshotDir(id)
+		renamedID = "rm-" + id
+		renamed = w.getSnapshotDir(renamedID)
+		if err = os.Rename(path, renamed); err != nil && !os.IsNotExist(err) {
+			if !os.IsPermission(err) {
+				return err
+			}
+			// If permission denied, it's possible that the scratch is still mounted, an
+			// artifact after a hard daemon crash for example. Worth a shot to try deactivating it
+			// before retrying the rename.
+			var (
+				home, layerID = filepath.Split(path)
+				di            = hcsshim.DriverInfo{
+					HomeDir: home,
+				}
+			)
+
+			if deactivateErr := hcsshim.DeactivateLayer(di, layerID); deactivateErr != nil {
+				return fmt.Errorf("failed to deactivate layer following failed rename: %s: %w", deactivateErr, err)
+			}
+
+			if renameErr := os.Rename(path, renamed); renameErr != nil && !os.IsNotExist(renameErr) {
+				return fmt.Errorf("second rename attempt following detach failed: %s: %w", renameErr, err)
+			}
+		}
+
+		restore = true
+		return nil
+	})
+	if err != nil {
+		if restore { // failed to commit
+			if err1 := os.Rename(renamed, path); err1 != nil {
+				// May cause inconsistent data on disk
+				log.G(ctx).WithError(err1).WithField("path", renamed).Error("Failed to rename after failed commit")
+			}
+		}
+		return "", err
+	}
+	return renamedID, nil
+}
+
+// Close closes the snapshotter
+func (w *windowsBaseSnapshotter) Close() error {
+	return w.ms.Close()
+}
+
+func (w *windowsBaseSnapshotter) createSnapshot(ctx context.Context, kind snapshots.Kind, key, parent string, opts []snapshots.Opt) (_ []mount.Mount, err error) {
+	var newSnapshot storage.Snapshot
+	err = w.ms.WithTransaction(ctx, true, func(ctx context.Context) error {
+		newSnapshot, err = storage.CreateSnapshot(ctx, kind, key, parent, opts...)
+		if err != nil {
+			return fmt.Errorf("failed to create snapshot: %w", err)
+		}
+
+		log.G(ctx).Debug("createSnapshot")
+		// Create the new snapshot dir
+		snDir := w.getSnapshotDir(newSnapshot.ID)
+		if err = os.MkdirAll(snDir, 0700); err != nil {
+			return fmt.Errorf("failed to create snapshot dir %s: %w", snDir, err)
+		}
+
+		if strings.Contains(key, snapshots.UnpackKeyPrefix) {
+			// IO/disk space optimization: Do nothing
+			//
+			// We only need one sandbox.vhdx for the container. Skip making one for this
+			// snapshot if this isn't the snapshot that just houses the final sandbox.vhd
+			// that will be mounted as the containers scratch. Currently the key for a snapshot
+			// where a layer will be extracted to will have the string `extract-` in it.
+			return nil
+		}
+
+		if len(newSnapshot.ParentIDs) == 0 {
+			// A parentless snapshot a new base layer. Valid base layers must have a "Files" folder.
+			// When committed, there'll be some post-processing to fill in the rest
+			// of the metadata.
+			filesDir := filepath.Join(snDir, "Files")
+			if err := os.MkdirAll(filesDir, 0700); err != nil {
+				return fmt.Errorf("creating Files dir: %w", err)
+			}
```

Comment:
Do we need `Files` dir for cimfs layers? What is it used for?

Reply:
At this point, I don't believe it can be a cimfs layer. This block _ought_ to only happen when creating a *writable* layer with no parents, which won't be cimfs format, it's just files-on-disk at this point, and we'll roll it into cimfs format on commit (by exporting to a tarstream and back like WCIFS, I assume).

I actually don't know if the containerd API lets you create a parentless View, but if it does (and we don't have any bugs...), you end up with a read-only symlink into the trivially empty Files directory, aka the WCIFS on-disk committed format, but using bindfs directly to the directory instead of to a WCIFS-created volume. Since Views can't be parents for future layers, there's no benefit I can see to trying to manage it with cimfs instead.

Reply:
Yeah since CimFS is only used for read-only images layers, the `if strings.Contains(key, snapshots.UnpackKeyPrefix) {` condition above would ensure we never reach here for CimFS layers.

Reply:
Got it. Thanks.